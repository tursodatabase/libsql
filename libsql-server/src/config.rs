use std::path::{Path, PathBuf};
use std::sync::Arc;

use anyhow::Context;
use hyper::client::HttpConnector;
use hyper_rustls::HttpsConnector;
use libsql_sys::EncryptionConfig;
use sha256::try_digest;
use tokio::time::Duration;
use tonic::transport::Channel;
use tower::ServiceExt;

use crate::auth::{Auth, Disabled};
use crate::net::{AddrIncoming, Connector};

pub struct RpcClientConfig<C = HttpConnector> {
    pub remote_url: String,
    pub tls_config: Option<TlsConfig>,
    pub connector: C,
}

impl<C: Connector> RpcClientConfig<C> {
    pub(crate) async fn configure(&self) -> anyhow::Result<(Channel, tonic::transport::Uri)> {
        let uri = tonic::transport::Uri::from_maybe_shared(self.remote_url.clone())?;
        let mut builder = Channel::builder(uri.clone());
        if let Some(ref tls_config) = self.tls_config {
            let cert_pem = std::fs::read_to_string(&tls_config.cert)?;
            let key_pem = std::fs::read_to_string(&tls_config.key)?;
            let identity = tonic::transport::Identity::from_pem(cert_pem, key_pem);

            let ca_cert_pem = std::fs::read_to_string(&tls_config.ca_cert)?;
            let ca_cert = tonic::transport::Certificate::from_pem(ca_cert_pem);

            let tls_config = tonic::transport::ClientTlsConfig::new()
                .identity(identity)
                .ca_certificate(ca_cert)
                .domain_name("sqld");
            builder = builder.tls_config(tls_config)?;
        }

        let channel =
            builder.connect_with_connector_lazy(self.connector.clone().map_err(Into::into));

        Ok((channel, uri))
    }
}

#[derive(Clone)]
pub struct TlsConfig {
    pub cert: PathBuf,
    pub key: PathBuf,
    pub ca_cert: PathBuf,
}

pub struct RpcServerConfig<A = AddrIncoming> {
    pub acceptor: A,
    pub tls_config: Option<TlsConfig>,
}

pub struct UserApiConfig<A = AddrIncoming> {
    pub hrana_ws_acceptor: Option<A>,
    pub http_acceptor: Option<A>,
    pub enable_http_console: bool,
    pub self_url: Option<String>,
    pub auth_strategy: Auth,
}

impl<A> Default for UserApiConfig<A> {
    fn default() -> Self {
        Self {
            hrana_ws_acceptor: Default::default(),
            http_acceptor: Default::default(),
            enable_http_console: Default::default(),
            self_url: Default::default(),
            auth_strategy: Auth::new(Disabled::new()),
        }
    }
}

pub struct AdminApiConfig<A = AddrIncoming, C = HttpsConnector<HttpConnector>> {
    pub acceptor: A,
    pub connector: C,
    pub disable_metrics: bool,
}

#[derive(Clone)]
pub struct DbConfig {
    pub extensions_path: Option<Arc<Path>>,
    pub bottomless_replication: Option<bottomless::replicator::Options>,
    pub max_log_size: u64,
    pub max_log_duration: Option<f32>,
    pub soft_heap_limit_mb: Option<usize>,
    pub hard_heap_limit_mb: Option<usize>,
    pub max_response_size: u64,
    pub max_total_response_size: u64,
    pub snapshot_exec: Option<String>,
    pub checkpoint_interval: Option<Duration>,
    pub snapshot_at_shutdown: bool,
    pub encryption_config: Option<EncryptionConfig>,
    pub max_concurrent_requests: u64,
}

impl Default for DbConfig {
    fn default() -> Self {
        Self {
            extensions_path: None,
            bottomless_replication: None,
            max_log_size: bytesize::mb(200u64),
            max_log_duration: None,
            soft_heap_limit_mb: None,
            hard_heap_limit_mb: None,
            max_response_size: bytesize::mb(10u64),
            max_total_response_size: bytesize::mb(10u64),
            snapshot_exec: None,
            checkpoint_interval: None,
            snapshot_at_shutdown: false,
            encryption_config: None,
            max_concurrent_requests: 128,
        }
    }
}

impl DbConfig {
    pub fn validate_extensions(&self) -> anyhow::Result<Arc<[PathBuf]>> {
        let mut valid_extensions = vec![];
        if let Some(ext_dir) = &self.extensions_path {
            let extensions_list = ext_dir.join("trusted.lst");

            let file_contents = std::fs::read_to_string(&extensions_list)
                .with_context(|| format!("can't read {}", &extensions_list.display()))?;

            let extensions = file_contents.lines().filter(|c| !c.is_empty());

            for line in extensions {
                let mut ext_info = line.trim().split_ascii_whitespace();

                let ext_sha = ext_info.next().ok_or_else(|| {
                    anyhow::anyhow!("invalid line on {}: {}", &extensions_list.display(), line)
                })?;
                let ext_fname = ext_info.next().ok_or_else(|| {
                    anyhow::anyhow!("invalid line on {}: {}", &extensions_list.display(), line)
                })?;

                anyhow::ensure!(
                    ext_info.next().is_none(),
                    "extension list seem to contain a filename with whitespaces. Rejected"
                );

                let extension_full_path = ext_dir.join(ext_fname);
                let digest = try_digest(extension_full_path.as_path()).with_context(|| {
                    format!(
                        "Failed to get sha256 digest, while trying to read {}",
                        extension_full_path.display()
                    )
                })?;

                anyhow::ensure!(
                    digest == ext_sha,
                    "sha256 differs for {}. Got {}",
                    ext_fname,
                    digest
                );
                valid_extensions.push(extension_full_path);
            }
        }

        Ok(valid_extensions.into())
    }
}

pub struct HeartbeatConfig {
    pub heartbeat_url: Option<String>,
    pub heartbeat_period: Duration,
    pub heartbeat_auth: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct MetaStoreConfig {
    pub bottomless: Option<BottomlessConfig>,
    pub allow_recover_from_fs: bool,
}

#[derive(Debug, Clone)]
pub struct BottomlessConfig {
    pub access_key_id: String,
    pub secret_access_key: String,
    pub session_token: Option<String>,
    pub region: String,
    pub backup_id: String,
    pub bucket_name: String,
    pub backup_interval: Duration,
    pub bucket_endpoint: String,
}
