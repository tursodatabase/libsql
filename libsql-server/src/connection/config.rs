use parking_lot::Mutex;
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::{fs, io};
use url::Url;

use crate::error::Error;
use crate::{Result, LIBSQL_PAGE_SIZE};

#[derive(Debug)]
pub struct DatabaseConfigStore {
    config_path: PathBuf,
    tmp_config_path: PathBuf,
    config: Mutex<Arc<DatabaseConfig>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatabaseConfig {
    #[serde(default)]
    pub block_reads: bool,
    #[serde(default)]
    pub block_writes: bool,
    /// The reason why operations are blocked. This will be included in [`Error::Blocked`].
    #[serde(default)]
    pub block_reason: Option<String>,
    /// maximum db size (in pages)
    #[serde(default = "default_max_size")]
    pub max_db_pages: u64,
    #[serde(default)]
    pub heartbeat_url: Option<Url>,
    #[serde(default)]
    pub bottomless_db_id: Option<String>,
}

const fn default_max_size() -> u64 {
    bytesize::ByteSize::pb(1000).as_u64() / LIBSQL_PAGE_SIZE
}

impl Default for DatabaseConfig {
    fn default() -> Self {
        Self {
            block_reads: Default::default(),
            block_writes: Default::default(),
            block_reason: Default::default(),
            max_db_pages: default_max_size(),
            heartbeat_url: None,
            bottomless_db_id: None,
        }
    }
}

impl DatabaseConfigStore {
    pub fn load(db_path: &Path) -> Result<Self> {
        let config_path = db_path.join("config.json");
        let tmp_config_path = db_path.join("config.json~");

        let config = match fs::read(&config_path) {
            Ok(data) => serde_json::from_slice(&data)?,
            Err(err) if err.kind() == io::ErrorKind::NotFound => DatabaseConfig::default(),
            Err(err) => return Err(Error::IOError(err)),
        };

        Ok(Self {
            config_path,
            tmp_config_path,
            config: Mutex::new(Arc::new(config)),
        })
    }

    #[cfg(test)]
    pub fn new_test() -> Self {
        Self {
            config_path: "".into(),
            tmp_config_path: "".into(),
            config: Mutex::new(Arc::new(DatabaseConfig::default())),
        }
    }

    pub fn get(&self) -> Arc<DatabaseConfig> {
        self.config.lock().clone()
    }

    pub fn store(&self, config: DatabaseConfig) -> Result<()> {
        let data = serde_json::to_vec_pretty(&config)?;
        fs::write(&self.tmp_config_path, data)?;
        fs::rename(&self.tmp_config_path, &self.config_path)?;
        *self.config.lock() = Arc::new(config);
        Ok(())
    }
}
