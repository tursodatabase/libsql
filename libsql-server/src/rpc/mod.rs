use std::sync::Arc;

use hyper_rustls::TlsAcceptor;
use libsql_replication::rpc::replication::NAMESPACE_METADATA_KEY;
use rustls::server::AllowAnyAuthenticatedClient;
use rustls::RootCertStore;
use tonic::Status;
use tower::util::option_layer;
use tower::ServiceBuilder;
use tower_http::trace::DefaultOnResponse;
use tracing::Span;

use crate::config::TlsConfig;
use crate::metrics::CLIENT_VERSION;
use crate::namespace::{NamespaceName, NamespaceStore};
use crate::rpc::proxy::rpc::proxy_server::ProxyServer;
use crate::rpc::proxy::ProxyService;
pub use crate::rpc::replication_log::rpc::replication_log_server::ReplicationLogServer;
use crate::rpc::replication_log::ReplicationLogService;
use crate::utils::services::idle_shutdown::IdleShutdownKicker;

pub mod proxy;
pub mod replica_proxy;
pub mod replication_log;
pub mod replication_log_proxy;
pub mod streaming_exec;

pub async fn run_rpc_server<A: crate::net::Accept>(
    proxy_service: ProxyService,
    acceptor: A,
    maybe_tls: Option<TlsConfig>,
    idle_shutdown_layer: Option<IdleShutdownKicker>,
    namespaces: NamespaceStore,
    disable_namespaces: bool,
) -> anyhow::Result<()> {
    let logger_service = ReplicationLogService::new(
        namespaces.clone(),
        idle_shutdown_layer.clone(),
        None,
        disable_namespaces,
        false,
    );

    if let Some(tls_config) = maybe_tls {
        let cert_pem = tokio::fs::read_to_string(&tls_config.cert).await?;
        let certs = rustls_pemfile::certs(&mut cert_pem.as_bytes())?;
        let certs = certs
            .into_iter()
            .map(rustls::Certificate)
            .collect::<Vec<_>>();

        let key_pem = tokio::fs::read_to_string(&tls_config.key).await?;
        let keys = rustls_pemfile::pkcs8_private_keys(&mut key_pem.as_bytes())?;
        let key = rustls::PrivateKey(keys[0].clone());

        let ca_cert_pem = std::fs::read_to_string(&tls_config.ca_cert)?;
        let ca_certs = rustls_pemfile::certs(&mut ca_cert_pem.as_bytes())?;
        let ca_certs = ca_certs
            .into_iter()
            .map(rustls::Certificate)
            .collect::<Vec<_>>();

        let mut roots = RootCertStore::empty();
        ca_certs.iter().try_for_each(|c| roots.add(c))?;
        let verifier = AllowAnyAuthenticatedClient::new(roots);
        let config = rustls::server::ServerConfig::builder()
            .with_safe_defaults()
            .with_client_cert_verifier(Arc::new(verifier))
            .with_single_cert(certs, key)?;

        let acceptor = TlsAcceptor::builder()
            .with_tls_config(config)
            .with_all_versions_alpn()
            .with_acceptor(acceptor);

        let router = tonic::transport::Server::builder()
            .layer(&option_layer(idle_shutdown_layer))
            .add_service(ProxyServer::new(proxy_service))
            .add_service(ReplicationLogServer::new(logger_service))
            .into_router();

        let svc = ServiceBuilder::new()
            .layer(
                tower_http::trace::TraceLayer::new_for_grpc()
                    .on_request(trace_request)
                    .on_response(
                        DefaultOnResponse::new()
                            .level(tracing::Level::DEBUG)
                            .latency_unit(tower_http::LatencyUnit::Micros),
                    ),
            )
            .service(router);

        tracing::info!("serving internal rpc server with tls");
        let h2c = crate::h2c::H2cMaker::new(svc);
        hyper::server::Server::builder(acceptor).serve(h2c).await?;
    } else {
        let proxy = ProxyServer::new(proxy_service);
        let replication = ReplicationLogServer::new(logger_service);

        let router = tonic::transport::Server::builder()
            .layer(&option_layer(idle_shutdown_layer))
            .add_service(proxy)
            .add_service(replication)
            .into_router();

        let svc = ServiceBuilder::new()
            .layer(
                tower_http::trace::TraceLayer::new_for_grpc()
                    .on_request(trace_request)
                    .on_response(
                        DefaultOnResponse::new()
                            .level(tracing::Level::DEBUG)
                            .latency_unit(tower_http::LatencyUnit::Micros),
                    ),
            )
            .service(router);

        let h2c = crate::h2c::H2cMaker::new(svc);

        tracing::info!("serving internal rpc server without tls");

        hyper::server::Server::builder(acceptor).serve(h2c).await?;
    }
    Ok(())
}

fn extract_namespace<T>(
    disable_namespaces: bool,
    req: &tonic::Request<T>,
) -> Result<NamespaceName, Status> {
    if disable_namespaces {
        return Ok(NamespaceName::default());
    }

    if let Some(namespace) = req.metadata().get_bin(NAMESPACE_METADATA_KEY) {
        let bytes = namespace
            .to_bytes()
            .map_err(|_| Status::invalid_argument("Metadata can't be converted into Bytes"))?;
        NamespaceName::from_bytes(bytes)
            .map_err(|_| Status::invalid_argument("Invalid namespace name"))
    } else {
        Err(Status::invalid_argument("Missing x-namespace-bin metadata"))
    }
}

fn trace_request<B>(req: &hyper::Request<B>, span: &Span) {
    let _s = span.enter();

    tracing::debug!(
        "rpc request: {} {} {:?}",
        req.method(),
        req.uri(),
        req.headers()
    );

    if let Some(v) = req.headers().get("x-libsql-client-version") {
        if let Ok(s) = v.to_str() {
            metrics::increment_counter!(CLIENT_VERSION, "version" => s.to_string());
        }
    }
}
