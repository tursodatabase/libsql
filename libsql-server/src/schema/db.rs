use libsql_replication::rpc::metadata;
use prost::Message;
use rusqlite::OptionalExtension;

use crate::connection::config::DatabaseConfig;
use crate::connection::program::Program;
use crate::namespace::NamespaceName;

use super::{
    status::{MigrationJob, MigrationTask},
    Error, MigrationJobStatus, MigrationTaskStatus,
};

pub(super) fn setup_schema(conn: &mut rusqlite::Connection) -> Result<(), Error> {
    conn.execute("PRAGMA foreign_key=ON", ())?;
    let txn = conn.transaction()?;

    // this table contains all the migration jobs
    txn.execute(
        &format!(
            "CREATE TABLE IF NOT EXISTS migration_jobs (
                job_id INTEGER PRIMARY KEY AUTOINCREMENT,
                schema TEXT NOT NULL,
                migration TEXT NOT NULL,
                status INTEGER,
                finished BOOLEAN GENERATED ALWAYS AS (status = {} OR status = {})
            )
            ",
            // TODO: also handle abort when we get there
            // we use format here, because params are not allowed GENERATED expression
            MigrationJobStatus::RunSuccess as u64,
            MigrationJobStatus::RunFailure as u64
        ),
        (),
    )?;
    // this table contains a list of all the that need to be performed for each migration job
    txn.execute(
        "CREATE TABLE IF NOT EXISTS migration_job_pending_tasks (
            task_id INTEGER PRIMARY KEY AUTOINCREMENT,
            job_id INTEGER,
            target_namespace TEXT NOT NULL,
            status INTEGER,
            error TEXT,
            FOREIGN KEY (job_id) REFERENCES migration_jobs (job_id)
        )
        ",
        (),
    )?;
    // This temporary table hold the list of tasks that are currently being processed
    txn.execute(
        "
        CREATE TEMPORARY TABLE IF NOT EXISTS enqueued_tasks (task_id)
        ",
        (),
    )?;

    // create a trigger that removes the task from enqueued tasks whenever it's status was updated.
    // The assumption is that the status of the task is only ever updated if work on it is
    // finished.
    txn.execute(
        "
        CREATE TEMPORARY TRIGGER IF NOT EXISTS remove_from_enqueued_tasks 
        AFTER UPDATE OF status ON migration_job_pending_tasks
        BEGIN
            DELETE FROM enqueued_tasks WHERE task_id = old.task_id;
        END
        ",
        (),
    )?;

    txn.commit()?;
    Ok(())
}

/// Create a migration job, and returns the job_id
pub(super) fn register_schema_migration_job(
    conn: &mut rusqlite::Connection,
    schema: &NamespaceName,
    migration: &Program,
) -> Result<i64, Error> {
    let txn = conn.transaction()?;

    // get the config for the schema and validate that it's actually a schema
    let mut stmt =
        txn.prepare("SELECT namespace, config FROM namespace_configs where namespace = ?")?;
    let mut rows = stmt.query([schema.as_str()])?;
    let Some(row) = rows.next()? else {
        return Err(Error::SchemaDoesntExist(schema.clone()));
    };
    let config_bytes = row.get_ref(1)?.as_blob().unwrap();
    // TODO: handle corrupted meta
    let config = DatabaseConfig::from(&metadata::DatabaseConfig::decode(config_bytes).unwrap());
    if !config.is_shared_schema {
        return Err(Error::NotASchema(schema.clone()));
    }

    drop(rows);

    stmt.finalize()?;

    let migration_serialized = serde_json::to_string(&migration).unwrap();
    // this query inserts the new job in migration_jobs only if there are no other unfinnished
    // tasks for that schema
    let row_changed = txn.execute("
        INSERT INTO migration_jobs (schema, migration, status)
        SELECT ?1, ?2, ?3 
        WHERE NOT (SELECT COUNT(1) FROM (SELECT 0 from migration_jobs WHERE schema = ?1 AND finished = false))
        ",
        (
            schema.as_str(),
            &migration_serialized,
            MigrationJobStatus::WaitingDryRun as u64,
        ),
    )?;

    if row_changed == 1 {
        let job_id = txn.last_insert_rowid();
        txn.execute(
            "
            INSERT INTO
            migration_job_pending_tasks (job_id, target_namespace, status)
            SELECT job_id, namespace, status
            FROM shared_schema_links 
            CROSS JOIN (SELECT ? as job_id, ? as status)
            WHERE shared_schema_name = ?",
            (
                job_id,
                MigrationTaskStatus::Enqueued as u64,
                schema.as_ref(),
            ),
        )?;

        txn.commit()?;
        Ok(job_id)
    } else {
        Err(Error::MigrationJobAlreadyInProgress(schema.clone()))
    }
}

/// returns a batch of tasks for job_id that are in the passed status
pub(super) fn get_next_pending_migration_tasks_batch(
    conn: &mut rusqlite::Connection,
    job_id: i64,
    status: MigrationTaskStatus,
    limit: usize,
) -> Result<Vec<MigrationTask>, Error> {
    let txn = conn.transaction_with_behavior(rusqlite::TransactionBehavior::Immediate)?;
    let tasks = txn
        .prepare(
            "SELECT task_id, target_namespace, status, job_id 
            FROM migration_job_pending_tasks 
            WHERE job_id = ? AND status = ? AND task_id NOT IN (select * from enqueued_tasks)
            LIMIT ?",
        )?
        .query_map((job_id, status as u64, limit), |row| {
            let task_id = row.get::<_, i64>(0)?;
            let namespace = NamespaceName::from_string(row.get::<_, String>(1)?).unwrap();
            let status = MigrationTaskStatus::from_int(row.get::<_, u64>(2)?);
            let job_id = row.get::<_, i64>(3)?;
            Ok(MigrationTask {
                namespace,
                status,
                job_id,
                task_id,
            })
        })?
        .map(|r| r.map_err(Into::into))
        .collect::<Result<Vec<_>, Error>>()?;

    for task in tasks.iter() {
        txn.execute("INSERT INTO enqueued_tasks VALUES (?)", [task.task_id])?;
    }

    txn.commit()?;
    Ok(tasks)
}

pub(super) fn update_meta_task_status(
    conn: &mut rusqlite::Connection,
    task: MigrationTask,
    error: Option<String>,
) -> Result<(), Error> {
    assert!(error.is_none() || task.status.is_failure());
    let txn = conn.transaction()?;
    txn.execute(
        "UPDATE migration_job_pending_tasks SET status = ?, error = ? WHERE task_id = ?",
        (task.status as u64, error, task.task_id),
    )?;
    txn.commit()?;
    Ok(())
}

/// Attempt to set the job to DryRunSuccess.
/// Checks that:
/// - current state is WaitinForDryRun
/// - all tasks are DryRunSuccess
pub(super) fn job_step_dry_run_success(
    conn: &mut rusqlite::Connection,
    mut job: MigrationJob,
) -> Result<MigrationJob, Error> {
    let row_changed = conn.execute(
        "
        WITH tasks AS (SELECT * FROM migration_job_pending_tasks WHERE job_id = ?1)
        UPDATE migration_jobs 
        SET status = ?2
        WHERE job_id = ?1
        AND status = ?3
        AND (SELECT count(1) from tasks) = (SELECT count(1) FROM tasks WHERE status = ?4)",
        (
            job.job_id(),
            MigrationJobStatus::DryRunSuccess as u64,
            MigrationJobStatus::WaitingDryRun as u64,
            MigrationTaskStatus::DryRunSuccess as u64,
        ),
    )?;

    if row_changed == 0 {
        return Ok(job);
    }

    *job.status_mut() = MigrationJobStatus::DryRunSuccess;

    Ok(job)
}

pub(super) fn update_job_status(
    conn: &mut rusqlite::Connection,
    job_id: i64,
    status: MigrationJobStatus,
) -> Result<(), Error> {
    conn.execute(
        "UPDATE migration_jobs SET status = ? WHERE job_id = ?",
        (status as u64, job_id),
    )?;
    Ok(())
}

pub(super) fn get_next_pending_migration_job(
    conn: &mut rusqlite::Connection,
) -> Result<Option<MigrationJob>, Error> {
    let txn = conn.transaction()?;
    let mut job = txn
        .query_row(
            "SELECT job_id, status, migration, schema
            FROM migration_jobs
            WHERE status != ? AND status != ?
            LIMIT 1",
            (
                MigrationJobStatus::RunSuccess as u64,
                MigrationJobStatus::RunFailure as u64,
            ),
            |row| {
                let job_id = row.get::<_, i64>(0)?;
                let status = MigrationJobStatus::from_int(row.get::<_, u64>(1)?);
                let migration = serde_json::from_str(row.get_ref(2)?.as_str()?).unwrap();
                let schema = NamespaceName::from_string(row.get::<_, String>(3)?).unwrap();
                Ok(MigrationJob {
                    schema,
                    job_id,
                    status,
                    migration,
                    progress: Default::default(),
                })
            },
        )
        .optional()?;

    if let Some(ref mut job) = job {
        txn.prepare(
            "
                SELECT status, count(1)
                FROM migration_job_pending_tasks 
                WHERE job_id = ?
                GROUP BY status",
        )?
        .query_map([job.job_id], |row| {
            job.progress[row.get::<_, usize>(0)?] = row.get::<_, usize>(1)?;
            Ok(())
        })?
        .collect::<Result<(), rusqlite::Error>>()?;
    }

    Ok(job)
}

#[cfg(test)]
mod test {
    use insta::assert_debug_snapshot;
    use tempfile::tempdir;

    use crate::namespace::meta_store::{metastore_connection_maker, MetaStore};

    use super::*;

    async fn register_schema(meta_store: &MetaStore, schema: &'static str) {
        meta_store
            .handle(schema.into())
            .store(DatabaseConfig {
                is_shared_schema: true,
                ..Default::default()
            })
            .await
            .unwrap();
    }

    async fn register_shared(meta_store: &MetaStore, name: &'static str, schema: &'static str) {
        meta_store
            .handle(name.into())
            .store(DatabaseConfig {
                shared_schema_name: Some(schema.into()),
                ..Default::default()
            })
            .await
            .unwrap();
    }

    #[tokio::test]
    async fn enqueue_migration_job() {
        let tmp = tempdir().unwrap();
        let (maker, manager) = metastore_connection_maker(None, tmp.path()).await.unwrap();
        let conn = maker().unwrap();
        let meta_store = MetaStore::new(Default::default(), tmp.path(), conn, manager)
            .await
            .unwrap();
        // create 2 shared schema tables
        register_schema(&meta_store, "schema1").await;
        register_schema(&meta_store, "schema2").await;

        // create namespaces
        register_shared(&meta_store, "ns1", "schema1").await;
        register_shared(&meta_store, "ns2", "schema2").await;
        register_shared(&meta_store, "ns3", "schema1").await;

        let mut conn = maker().unwrap();
        setup_schema(&mut conn).unwrap();
        register_schema_migration_job(
            &mut conn,
            &"schema1".into(),
            &Program::seq(&["select * from test"]),
        )
        .unwrap();

        let mut stmt = conn.prepare("select * from migration_jobs").unwrap();
        assert_debug_snapshot!(stmt.query(()).unwrap().next().unwrap().unwrap());
        stmt.finalize().unwrap();

        let mut stmt = conn
            .prepare("select * from migration_job_pending_tasks")
            .unwrap();
        let mut rows = stmt.query(()).unwrap();
        assert_debug_snapshot!(rows.next().unwrap().unwrap());
        assert_debug_snapshot!(rows.next().unwrap().unwrap());
        assert!(rows.next().unwrap().is_none());
    }

    #[tokio::test]
    async fn schema_doesnt_exist() {
        let tmp = tempdir().unwrap();
        let (maker, manager) = metastore_connection_maker(None, tmp.path()).await.unwrap();
        let conn = maker().unwrap();
        let meta_store = MetaStore::new(Default::default(), tmp.path(), conn, manager)
            .await
            .unwrap();

        // FIXME: the actual error reported here is a shitty constraint error, we should make the
        // necessary checks beforehand, and return a nice error message.
        assert!(meta_store
            .handle("ns1".into())
            .store(DatabaseConfig {
                shared_schema_name: Some("schema1".into()),
                ..Default::default()
            })
            .await
            .is_err());
    }

    #[tokio::test]
    async fn pending_job() {
        let tmp = tempdir().unwrap();
        let (maker, manager) = metastore_connection_maker(None, tmp.path()).await.unwrap();
        let conn = maker().unwrap();
        let meta_store = MetaStore::new(Default::default(), tmp.path(), conn, manager)
            .await
            .unwrap();

        register_schema(&meta_store, "schema1").await;
        register_shared(&meta_store, "ns1", "schema1").await;
        register_shared(&meta_store, "ns2", "schema1").await;
        register_shared(&meta_store, "ns3", "schema1").await;

        let mut conn = maker().unwrap();
        setup_schema(&mut conn).unwrap();

        let job_id = register_schema_migration_job(
            &mut conn,
            &"schema1".into(),
            &Program::seq(&["create table test (x)"]).into(),
        )
        .unwrap();

        assert_debug_snapshot!(get_next_pending_migration_job(&mut conn).unwrap().unwrap());

        let mut tasks = get_next_pending_migration_tasks_batch(
            &mut conn,
            job_id,
            MigrationTaskStatus::Enqueued,
            10,
        )
        .unwrap();
        assert_debug_snapshot!(tasks);

        let mut task = tasks.pop().unwrap();
        *task.status_mut() = MigrationTaskStatus::Success;
        update_meta_task_status(&mut conn, task, None).unwrap();

        assert_debug_snapshot!(get_next_pending_migration_job(&mut conn).unwrap().unwrap());
    }

    #[tokio::test]
    async fn step_job_dry_run_success() {
        let tmp = tempdir().unwrap();
        let (maker, manager) = metastore_connection_maker(None, tmp.path()).await.unwrap();
        let conn = maker().unwrap();
        let meta_store = MetaStore::new(Default::default(), tmp.path(), conn, manager)
            .await
            .unwrap();

        register_schema(&meta_store, "schema1").await;
        register_shared(&meta_store, "ns1", "schema1").await;
        register_shared(&meta_store, "ns2", "schema1").await;
        register_shared(&meta_store, "ns3", "schema1").await;

        let mut conn = maker().unwrap();
        setup_schema(&mut conn).unwrap();
        register_schema_migration_job(
            &mut conn,
            &"schema1".into(),
            &Program::seq(&["create table test (x)"]).into(),
        )
        .unwrap();

        let job = get_next_pending_migration_job(&mut conn).unwrap().unwrap();
        let job = job_step_dry_run_success(&mut conn, job).unwrap();

        // the job status wasn't updated: there are still tasks that need dry run
        assert_eq!(*job.status(), MigrationJobStatus::WaitingDryRun);

        let tasks = get_next_pending_migration_tasks_batch(
            &mut conn,
            job.job_id(),
            MigrationTaskStatus::Enqueued,
            10,
        )
        .unwrap();
        for mut task in tasks {
            task.status = MigrationTaskStatus::DryRunSuccess;
            update_meta_task_status(&mut conn, task, None).unwrap();
        }

        let job = job_step_dry_run_success(&mut conn, job).unwrap();
        assert_eq!(job.status, MigrationJobStatus::DryRunSuccess);
    }

    #[tokio::test]
    async fn cannot_enqueue_another_job_for_namespace_while_other_job_still_pending() {
        let tmp = tempdir().unwrap();
        let (maker, manager) = metastore_connection_maker(None, tmp.path()).await.unwrap();
        let conn = maker().unwrap();
        let meta_store = MetaStore::new(Default::default(), tmp.path(), conn, manager)
            .await
            .unwrap();

        register_schema(&meta_store, "schema1").await;
        register_schema(&meta_store, "schema2").await;

        let mut conn = maker().unwrap();
        setup_schema(&mut conn).unwrap();

        let job_id = register_schema_migration_job(
            &mut conn,
            &"schema1".into(),
            &Program::seq(&["create table test (x)"]).into(),
        )
        .unwrap();

        // cannot create a job for a task that has a pending job
        assert!(matches!(
            register_schema_migration_job(
                &mut conn,
                &"schema1".into(),
                &Program::seq(&["create table test (x)"]).into(),
            )
            .unwrap_err(),
            Error::MigrationJobAlreadyInProgress(_)
        ));

        // ok for another schema without pending job
        register_schema_migration_job(
            &mut conn,
            &"schema2".into(),
            &Program::seq(&["create table test (x)"]).into(),
        )
        .unwrap();

        update_job_status(&mut conn, job_id, MigrationJobStatus::RunSuccess).unwrap();

        // job is finished, we can enqueue now
        register_schema_migration_job(
            &mut conn,
            &"schema1".into(),
            &Program::seq(&["create table test (x)"]).into(),
        )
        .unwrap();
    }
}
