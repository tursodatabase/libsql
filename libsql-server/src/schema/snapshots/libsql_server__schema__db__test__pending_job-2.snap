---
source: libsql-server/src/schema/db.rs
expression: tasks
---
[
    MigrationTask {
        namespace: ns1,
        status: Enqueued,
        job_id: 1,
        task_id: 1,
    },
    MigrationTask {
        namespace: ns2,
        status: Enqueued,
        job_id: 1,
        task_id: 2,
    },
    MigrationTask {
        namespace: ns3,
        status: Enqueued,
        job_id: 1,
        task_id: 3,
    },
]
